cmd/kops/update_cluster.go

```diff
@@ -138,6 +143,22 @@ func RunUpdateCluster(ctx context.Context, f *util.Factory, clusterName string,
 	isDryrun := false
 	targetName := c.Target
 
+	if c.admin && c.user != "" {
+		return nil, fmt.Errorf("cannot use both --admin and --user")
+	}
+
+	if c.CreateKubecfg && !c.admin && c.user == "" {
+		return nil, fmt.Errorf("--create-kube-config requires that either --admin or --user is set")
```

Comment:
To some extent code related. It is not straightforward in the current code to see if there is already a context for the cluster early in the `update` code. I didn't want to let the user run `update` to provision a new cluster, only for the command to fail at the very end when exporting the user and _then_ complain about `--admin` or `--user` not being set.

So we assume that the user only adds if they want to add the new context `--create-kube-config`. That is also why I defaulted this flag to false rather than true now since usually one only wants this if one runs `update` to create the cloud resources the first time.

The other use case I can think of is when one rotates certs, and at least if one uses the admin user, one would need to add `--admin` anyway.

Reply:
I believe it makes sense to export a context for the cluster without either adding admin credentials or setting the context to a precreated singular user.

Our authentication setup has per-cluster tokens which are obtained by an exec hook. That means a single user shared across multiple clusters would not work.

Reply:
I am a bit concerned that if one sets `--create-kube-config`, but does not catch that one _probably should_ also set `--admin` or `--user` one will get a broken config. So I do agree that there are use cases for not doing this, I am wondering if this is worth it usabilitywise.

For your use case, you have to remove the user from the context once and then kops will at least leave the context alone and not keep re-adding the admin user, which is a good benefit on its own I should think. I assume that typically you would only export the cluster config once per operator per cluster so this isn't too much hassle.

Reply:
Since `--admin` and `--user` imply `--create-kube-config` it is unlikely that `--create-kube-config` would be explicitly set unless the invoker is doing something special.

Our wrapper around kops has two forms of `export kubecfg`, one which corresponds to `--admin` and one which adds a per-cluster user with our local exec hook added. Since the admin credential is so powerful, I would prefer not to have that second case extract the credential, write it to disk, and then try to remove it.

I suppose I could let the current PR through and then address this case in a followup PR. I am planning on reducing the default lifetime of the admin credential to 18 hours.

Reply:
I think that would be helpful. You have a much better idea of how this should work for your use case than I do.

I do agree that passing only `--create-kube-config` would be something special. If you log a warning about creating a context without user, I think that should be fine. It would at least give a hint to the user about why the credentials aren't working.